content/courses/advanced-ada/chapters/access_types.rst

```diff
@@ -294,14 +294,402 @@ Unchecked Access
+    end Show_Unchecked_Deallocation;
+
+As we've seen before, we can have multiple calls to :ada:`Free (I)`. However,
+the call to :ada:`I_2` fails because it points to an object that doesn't exist
+anymore.
+
+Because of these potential errors, it is the programmer's responsibility to be
+very careful when using unchecked deallocation and avoid creating dangling
+references.
+
+For the example we've just seen, a better approach could be to explicitly
+assign :ada:`null` to :ada:`I_2` to indicate that it doesn't point to any
+specific object:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Access_Types.Unchecked_Deallocation
+
+    with Integer_Types; use Integer_Types;
+
+    procedure Show_Unchecked_Deallocation is
+
+       I, I_2 : Integer_Access;
+
+    begin
+       I := new Integer;
+
+       I_2 := I;
+
+       --  NOTE: I_2 points to the same
+       --        object as I.
+
+       Free (I);
+
+       I_2 := null;
+
+       --  NOTE: I_2 doesn't point to any
+       --        object, so we can call Free.
+
+       Free (I_2);
+    end Show_Unchecked_Deallocation;
+
+Now, calling :ada:`Free (I_2)` doesn't cause any issues because it doesn't
+point to any object.
+
+
+Dereferencing dangling references
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Of course, we shouldn't try to dereference a dangling reference. For example:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Access_Types.Unchecked_Deallocation
+
+    with Ada.Text_IO;   use Ada.Text_IO;
+    with Integer_Types; use Integer_Types;
+
+    procedure Show_Unchecked_Deallocation is
+
+       I_1, I_2 : Integer_Access;
+
+    begin
+       I_1 := new Integer'(42);
+       I_2 := I_1;
+
+       Put_Line ("I_1 = " & Integer'Image (I_1.all));
+       Put_Line ("I_2 = " & Integer'Image (I_2.all));
+
+       Put_Line ("Freeing I_1");
+       Free (I_1);
+
+       if I_1 /= null then
+          Put_Line ("I_1 = " & Integer'Image (I_1.all));
+       end if;
+
+       if I_2 /= null then
+          Put_Line ("I_2 = " & Integer'Image (I_2.all));
+       end if;
+    end Show_Unchecked_Deallocation;
+
+In this example, we allocate an object for :ada:`I_1` and make :ada:`I_2` point
+to the same object. Note, however, that we call :ada:`Free` for :ada:`I_1`, but
+not for :ada:`I_2`. Therefore, after that call to :ada:`Free`,
+:ada:`I_1 = null`, but :ada:`I_2` still points to the object originally
+allocated for :ada:`I_1`.
+
+As mentioned previously, when deallocating an object via a call to :ada:`Free`,
+the corresponding access value is set to :ada:`null`. In this case, we can
+check for :ada:`null` to detect a dangling reference and avoid dereferencing
+it.
+
+The strategy of checking for :ada:`null` before trying to dereferencing the
+access value works fine for :ada:`I_1`. However, this strategy doesn't work for
+:ada:`I_2`: the access value is not :ada:`null` because it still points to an
+object that existed at some point. Therefore, the application will try to
```

Comment:
Yes, exactly right, so what I am suggesting is careful wording in the earlier text.